modules/metrics/src/main/java/org/motechproject/metrics/service/impl/MetricRegistryServiceImpl.java

```diff
@@ -0,0 +1,204 @@
+    private final MetricsConfigFacade metricsConfigFacade;
+
+    private Map<String, Metric> metrics;
+
+    @Autowired
+    public MetricRegistryServiceImpl(MetricRegistry metricRegistry, MetricsConfigFacade metricsConfigFacade) {
+        this.metricRegistry = metricRegistry;
+        this.metricsConfigFacade = metricsConfigFacade;
+        metrics = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void setEnabled(boolean enabled) {
+        for (Metric metric: metrics.values()) {
+            if (metric instanceof Enablable && ((Enablable) metric).isEnabled() != enabled) {
+                ((Enablable) metric).setEnabled(enabled);
+            }
+        }
+    }
+
+    @Override
+    public Counter counter(final String name) {
+        return getOrAdd(name, MetricBuilder.COUNTERS);
+    }
+
+    @Override
+    public Histogram histogram(final String name) {
+        return getOrAdd(name, MetricBuilder.HISTOGRAMS);
+    }
+
+    @Override
+    public Meter meter(final String name) {
+        return getOrAdd(name, MetricBuilder.METERS);
+    }
+
+    @Override
+    public Timer timer(final String name) {
+        return getOrAdd(name, MetricBuilder.TIMERS);
+    }
+
+    @Override
+    public <T> Gauge<T> registerGauge(final String name, final Gauge<T> gauge) {
+        com.codahale.metrics.Gauge<T> theGauge;
+        try {
+            theGauge = metricRegistry.register(name, new com.codahale.metrics.Gauge<T>() {
+                @Override
+                public T getValue() {
+                    return gauge.getValue();
+                }
+            });
+        } catch (IllegalArgumentException ex) {
+            throw ex;
+        }
+        return new GaugeAdapter<T>(theGauge);
+    }
+
+    @Override
+    public <T extends Number> Gauge<Double> registerRatioGauge(final String name, Supplier<T> numerator, Supplier<T> denominator) {
+        com.codahale.metrics.RatioGauge theGauge;
+        try {
+            theGauge = metricRegistry.register(name, new RatioGauge() {
+                @Override
+                protected Ratio getRatio() {
+                    return Ratio.of(numerator.get().doubleValue(), denominator.get().doubleValue());
+                }
+            });
+        } catch (IllegalArgumentException ex) {
+            throw ex;
+        }
+        return new GaugeAdapter<>(theGauge);
+    }
+
+    @Override
+    public boolean isRegistered(String name) {
+        return metricRegistry.getNames().contains(name);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T extends Metric> T getOrAdd(String name, MetricBuilder<T> builder) {
+        Metric metric = metrics.get(name);
+
+        if (metric == null) {
+            try {
+                T built = builder.createMetric(name, metricRegistry, metricsConfigFacade);
+                metrics.put(name, built);
+                return built;
+            } catch (IllegalArgumentException ex) {
+                throw ex;
```

Comment:
Doesn't make sense to catch the exception only to throw the same exception again. Did you mean to wrap it into custom exception?